Sources/Sentry/SentryANRTracker.m

```diff
@@ -47,60 +48,34 @@ - (instancetype)initWithTimeoutInterval:(NSTimeInterval)timeoutInterval
 - (void)detectANRs
 {
     NSThread.currentThread.name = @"io.sentry.app-hang-tracker";
-
     self.thread = NSThread.currentThread;
 
-    BOOL wasPreviousANR = NO;
+    __block NSInteger ticksSinceUiUpdate = 0;
+    NSInteger reportTreshold = 5;
+    NSTimeInterval sleepInterval = self.timeoutInterval / reportTreshold;
 
     while (![self.thread isCancelled]) {
+        ticksSinceUiUpdate++;
 
-        NSDate *blockDeadline =
-            [[self.currentDate date] dateByAddingTimeInterval:self.timeoutInterval];
+        [self.dispatchQueueWrapper dispatchOnMainQueue:^{
+            ticksSinceUiUpdate = 0;
+            self->reported = NO;
+        }];
 
-        __block BOOL blockExecutedOnMainThread = NO;
-        [self.dispatchQueueWrapper dispatchOnMainQueue:^{ blockExecutedOnMainThread = YES; }];
+        [self.threadWrapper sleepForTimeInterval:sleepInterval];
 
-        [self.threadWrapper sleepForTimeInterval:self.timeoutInterval];
+        if (ticksSinceUiUpdate >= reportTreshold && !reported) {
+            self->reported = YES;
 
-        if (blockExecutedOnMainThread) {
-            if (wasPreviousANR) {
-                [SentryLog logWithMessage:@"ANR stopped." andLevel:kSentryLevelWarning];
-                [self ANRStopped];
```

Comment:
h: you are missing this in the new approach.
End of ANR indication

Reply:
Added it back, in a slightly different way. I think it should work 🤞

Reply:
Not sure about running it in the main thread but I think it works.